cabot/metricsapp/models/elastic.py

```diff
@@ -37,3 +49,132 @@ def client(self):
         if self._client:
             return self._client
         return create_es_client(self.urls, self.timeout)
+
+
+class ElasticsearchStatusCheck(MetricsStatusCheckBase):
+    class Meta:
+        app_label = 'metricsapp'
+
+    queries = models.TextField(
+        max_length=10000,
+        help_text='List of raw json Elasticsearch queries. Format: [q] or [q1, q2, ...]. '
+                  'Query guidelines: all aggregations should be named "agg." The most '
+                  'internal aggregation must be a date_histogram. Metrics names should be '
+                  'the same as the metric types (e.g., "max", "min", "avg").'
+    )
+
+    def get_series(self):
+        """
+        Get the relevant data for a check from Elasticsearch and parse it
+        into a generic format.
+        :param check: the ElasticsearchStatusCheck
+        :return data in the format
+            status:
+            error_message:
+            error_code:
+            raw:
+            data:
+              - series: a.b.c.d
+                datapoints:
+                  - [timestamp, value]
+                  - [timestamp, value]
+              - series: a.b.c.p.q
+                datapoints:
+                  - [timestamp, value]
+                check:
+        """
+        parsed_data = dict()
+        parsed_data['raw'] = None
+        # Will be set to true if we encounter an error
+        parsed_data['error'] = False
+        parsed_data['data'] = []
+
+        try:
+            queries = json.loads(self.queries)
+        except ValueError as e:
+            logger.exception('Error loading Elasticsearch queries: {}'.format(self.queries))
+            parsed_data['error_code'] = type(e).__name__
+            parsed_data['error_message'] = str(e)
+            parsed_data['error'] = True
+            return parsed_data
+
+        for query in queries:
+            source = ElasticsearchSource.objects.get(name=self.source.name)
+            try:
+                search = Search().from_dict(query)
+                response = search.using(source.client).index(source.index).execute()
+                parsed_data['raw'] = response.to_dict()
+                parsed_data['data'].extend(self._es_rec([response.to_dict()['aggregations']]))
+            except Exception as e:
+                logger.exception('Error executing Elasticsearch query: {}'.format(query))
+                parsed_data['error_code'] = type(e).__name__
+                parsed_data['error_message'] = str(e)
+                parsed_data['error'] = True
+                break
+
+        return parsed_data
+
+    def _es_rec(self, series, series_name=None):
```

Comment:
So, I have been thinking about ways to simplify this function. How does this look?

```python
import json
from collections import defaultdict


class Boo(object):
    def _parse_es_response(self, filename):
        earliest_point = 0
        output = []
        series = [json.load(open(filename))['aggregations']]

        data = defaultdict(list)
        for metric, (ts, value) in self._parse_series(series):
            if ts > earliest_point:
                data[metric].append((ts, value))

        for series, datapoints in data.iteritems():
            output.append(dict(series=series, datapoints=datapoints))

        return output

    def _parse_series(self, series, series_name=None):
        """
        Look through the json response recursively to go through all the
        aggregation buckets
        :param series: the aggregations part of the ES response
        :param series_name: the name of the series we're looking at
            ("key1.key2....metric_name")
        """

        # All aggregations should be named 'agg'
        for subseries in series:
            # If there are no more aggregations, we've reached the metric
            if subseries.get('agg') is None:
                # Note: I am sending the subseries inside
                results = self._get_metric_data(subseries, series_name)

            else:
                # New name is "series_name.subseries_name" (if they exist)
                key = subseries.get('key', 'metric')
                sub_series_name = '.'.join(filter(None, [series_name, key]))
                results = self._parse_series(subseries['agg']['buckets'],
                                             series_name=sub_series_name)

            for result in results:
                yield result

    def _get_metric_data(self, subseries, series_name):
        """
        Extract the metric name and values from a json series
        :param series: subset of the ES response that includes a list of
            metric values with date histogram keys
        :param series_name: the name of the series we're looking at
        """
        ts = subseries['key'] / 1000

        for metric_name, detail in subseries.iteritems():
            if type(detail) != dict:
                continue

            metric_name = '{}.{}'.format(series_name, metric_name)

            if 'value' in detail:
                yield (metric_name, (ts, detail['value']))

            elif 'values' in detail:
                for key, value in detail['values'].iteritems():
                    sub_metric_name = '{}.{}'.format(metric_name, key)
                    yield (sub_metric_name, (ts, value))

            else:
                raise NotImplementedError()


if __name__ == '__main__':
    import sys
    b = Boo()
    print(json.dumps(b._parse_es_response(sys.argv[1]), indent=1))
```

How does it work? Generators. The entire parsing operation uses generators. Advantage: we handle the data very lazily. More memory efficient. The parsing into output and filtering is done at the very end.

Also, do you think that the code looks simpler?